### ISSUE TITLE
OAuth2 and SSE Notificiations 

### ISSUE THREAD

**Issue by farshidtz** [w3c/wot-profile#397]

**Comment by benfrancis** [comment]
> Security TF discussed, felt this was not a discovery topic per se, since it applies to regular Things as well. Probably should be covered elsewhere, perhaps in an SSE Profile, where it makes sense to recommend "best practices".
> Rather than closing this issue though I think it should be "moved", e.g. to Profiles.

I agree that what (if anything) to mandate or recommend about token expiry and renewal is an important topic in general (see also #347), and SSE is indeed an interesting special case due to the extended duration of open connections and limitations of the client-side API in browsers.

However, note that recommending or mandating best practices in a profile will not also solve this problem for Directory implementations, which may not necessarily conform to a profile. I'm not sure if the Directory Service API is even _compatible_ with the existing profiles (the Directory Service API is very much like a profile for Directories itself, which isn't necessarily consistent with the constraints in the WoT Profiles specification).

So yes, this may be something that can be addressed in profiles for Things in general, but in order to solve this problem for Directories we'd need to either:
- Separately specify the same constraints in the WoT Discovery specification (as per the original issue)
- Add support for Directories in Profiles, either via a dedicated Directory profile(s), or by including support for Directories in the existing profiles (see https://github.com/w3c/wot-profile/issues/112#issuecomment-1020144745)

Also note that if these really are just recommended "best practices" which apply to all Things (as opposed to mandatory constraints needed to guarantee interoperability between Things/Consumers conforming to a profile), I would suggest they would be better defined in the Thing Description specification, or Security Best Practices document.

---

For the record, WebThings Gateway currently includes JWTs in a query string for SSE (and WebSocket) connections, since that's basically the only option for the native EventSource (and WebSocket) APIs in web browsers (it's theoretically possible to use cookies, but only via a caching mechanism which is not useful in practice). The tokens also do not currently have an expiry date which is obviously not ideal, so I would like to properly implement token expiry and renewal as part of OAuth2 for the gateway's REST API. I don't really know how to solve that for SSE and WebSockets without making it impossible to implement using the standardised browser APIs.
